Tests/AsyncQueueTests/ActorQueueTests.swift

```diff
@@ -0,0 +1,184 @@
+                }
+            }
+            var unsafeCount = 0
+            let dispatchQueue = DispatchQueue(label: #function)
+        }
+        let counter = SynchronousCounter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                counter.incrementAndExpectCount(equals: iteration)
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_async_startsExecutionOfNextTaskAfterSuspension() async {
+        let counter = Counter()
+        let semaphore = Semaphore()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: iteration)
+                await semaphore.wait()
+            }
+        }
+        systemUnderTest.async {
+            for _ in 1...1_000 {
+                await semaphore.signal()
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_async_allowsReentrancy() async {
+        let counter = Counter()
+        await systemUnderTest.await { [systemUnderTest] in
+            await systemUnderTest.await {
+                await counter.incrementAndExpectCount(equals: 1)
+            }
+            await counter.incrementAndExpectCount(equals: 2)
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    @MainActor
+    func test_async_doesNotRetainTaskAfterExecution() async {
+        let expectation = self.expectation(description: #function)
+        final class Reference: Sendable {}
+        final class ReferenceHolder: @unchecked Sendable {
+            var reference: Reference? = Reference()
+        }
+        let referenceHolder = ReferenceHolder()
+        weak var weakReference = referenceHolder.reference
+        let asyncSemaphore = Semaphore()
+        let syncSemaphore = Semaphore()
+        systemUnderTest.async { [reference = referenceHolder.reference] in
+            // Now that we've started the task and captured the reference, release the synchronous code.
+            await syncSemaphore.signal()
+            // Wait for the synchronous setup to complete and the reference to be nil'd out.
+            await asyncSemaphore.wait()
+            // Retain the unsafe counter until the task is completed.
+            _ = reference
+            // Signal that this task is about to clean up.
+            expectation.fulfill()
+        }
+        // Wait for the asynchronous task to start.
+        await syncSemaphore.wait()
+        referenceHolder.reference = nil
+        XCTAssertNotNil(weakReference)
+        // Allow the enqueued task to complete.
+        await asyncSemaphore.signal()
+        // Make sure the task has completed.
+        waitForExpectations(timeout: 1.0)
+        // Wait a runloop to ensure the previously enqueued task has had time to deallocate.
+        DispatchQueue.main.async {
+            XCTAssertNil(weakReference)
+        }
```

Comment:
We didn't need this code to wait a runloop in the `FIFOQueue` tests. Why do we need it here?

Reply:
Also, by dispatching async on the main queue the assertion runs after this method completes. If that's the case and I'm not missing something it's something we should avoid. Can we use an `@autoreleasepool` instead?

Reply:
If we do need to use an async onto the main queue I think we want to write an expectation that we fulfill after the async code completes, and then wait for the expectation at the end of the test. That ensures that we always execute the `XCTAssertNil` before `test_async_doesNotRetainTaskAfterExecution` returns.

Reply:
> I tried writing this test with an `autoreleasepool` those do not work with `async` closures.

Would this approach help us? https://stackoverflow.com/a/68976711/280789 If so this feels a bit more explicit about what we're trying to achieve wrt memory management.

Reply:
The issue is less that ARC hasn't finished yet as much as it is that **guaranteeing** that we execute _after_ the block has completed execution is difficult. I don't think an autorelease pool will help us much, especially one wrapped in a `Task` because controlling when a `Task` executes is seriously difficult.

Reply:
> If we do need to use an async onto the main queue I think we want to write an expectation that we fulfill after the async code completes, and then wait for the expectation at the end of the test. That ensures that we always execute the `XCTAssertNil` before `test_async_doesNotRetainTaskAfterExecution` returns.

Good point that we might execute the `XCTAssertNil` too late though. That's hopefully an easy fix with an `await` and another expectation.

Reply:
Simplified dramatically in b313127. The key was that I remembered that the actor queue is still executing tasks atomically _as long as those tasks do not await anything_.